WordPress/openverse
# Refactor the media collection

## Fixes



Fixes #3576 by @AetherUnbound
Fixes #3830 by @obulat 

_Note: this PR does not fix the problem when the creator/tag have special symbols (`/` or `?`)_

## Description


Currently, we fetch the search results in two separate places. The `search.vue` component calls `fetchMedia` when the search parameters like the search type, search term or filters change. The `VLoadMore` component calls `fetchMedia` when the user clicks on "Load more" button, fetching the pages after the first one.

This means that we cannot use the list of media returned in `search.vue`'s `fetchMedia` call, because it won't be updated when the "Load more" button is clicked, and only the first page will be shown.
To work around this limitation, we were using the computed that returned the value of media results from the media store.  These values were passed to the image/audio search pages, but the All content page would use a separate computed.

This PR tries to separate the concerns more cleanly.

### Data fetching
The **pages** (`search.vue`, `_tag.vue`, `_.vue` for source and creator views) are responsible for **fetching the data**, including the load more fetching (that was previously done by the `VLoadMore` component). 

### Data display
The `VMediaCollection` is responsible for displaying the media search results or the loading skeleton.

It handles both main and additional search views, and displays the header with the title and link buttons, the media grid, the load more button and the external search form (if necessary) and the scroll to top button.

### Updating the data on navigation

The linked issue is that only the first page of additional search views results was kept if you navigate to the single result page from the second page (or third, fourth, etc) and back. To fix it, we set the media to the media results from the store if the `collectionParams` match the path parameters. 
This fix created a different issue: the old media results were shown when you would navigate between the single result views and the additional search results views. For instance, when going back from the single results views, or when navigating between the views (e.g., tag view -> single result -> creator view -> single result -> source view and back). To fix that, we clear the media in the `media` store if the `collectionParams` parsed from the path do not match the one in the search store.

## Testing Instructions

Make sure the `additional_search_views` feature flag is `on` (use the switch on the `/preferences` page)
Go to a tag view (localhost:8443/image/tag/cat), load a couple more pages of results. Then select one of the results from the last page. Go back by clicking "Back to results" or the browser's "Back" button. You should see that all of the previously loaded results are displayed (instead of the first page as in current implementation).

Also try to navigate between the additional search views and the single result views (e.g. tag/cat -> single result -> creator view -> single result -> source view and back), and check that the title of the additional search view always matches the results. You should never see, for example, results for a creator and title of "tag Cat".

Repeat the steps for #3830 and check that it's fixed.

## Checklist



[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
